First version of developer overview (#537)
NVIDIA/cub: docs/developer_overview.md

# CUB Developer Overview

> As a CUB developer, you might experience difficulties reconstructing CUB design by reading source codes. This document is written to ease this reconstruction. At the high level, you can notice that CUB has a layered structure. Each layer (scope) corresponds to a particular hardware level:

Reviewer: ```suggestion

This living document serves as a guide to the design of the internal structure of CUB. 

The algorithms CUB provides are layered in a way that corresponds to the hierarchy of threads in CUDA where higher-level layers build on top of lower. 

These layers are:
- Thread 
- Warp
- Block
- Device

For example, consider `reduce`. CUB has four flavors of `reduce` that correspond to these different layers:

```

# CUB Developer Overview

This living document serves as a guide to the design of the internal structure of CUB. 

The algorithms CUB provides are layered in a way that corresponds to the hierarchy of threads in CUDA where higher-level layers build on top of lower. 

These layers are:
- Thread 
- Warp
- Block
- Device

> For example, consider `reduce`. CUB has four flavors of `reduce` that correspond to these different layers:
> 

Reviewer: This may be walking the line between developer and user guide. We can evolve this doc over time once we actually have both a user and dev guide :slightly_smiling_face:. 

```suggestion
For example, consider `reduce`. CUB has four flavors of `reduce` that correspond to these different layers. Each layer has unique semantics with respect to how they are invoked, how many threads participate, and on which thread(s) the result is valid. 

`cub::ThreadReduce` 
   - A normal function invoked and executed sequentially by a single thread that returns a valid result on that thread

`cub::WarpReduce` and `cub::BlockReduce`
   - A "cooperative" function where threads concurrently invoke the same function to execute parallel work
   - The function's return value is well-defined only on the "first" thread (lowest thread index)

`cub::DeviceReduce` 
   - A normal function invoked by a single thread that spawns additional threads to execute parallel work
   - Result is stored in the pointer provided to the function
   - Function returns a `cudaError_t` error code

The table below provides a summary of these functions:
```

| layer                | coop invocation | coop execution | max threads | valid result in |
| -------------------- | --------------- | -------------- | ----------- | --------------- |
| `cub::ThreadReduce`  | -               | -              | 1           | invoking thread |
| `cub::WarpReduce`    | +               | +              | 32          | main thread     |
| `cub::BlockReduce`   | +               | +              | 1024        | main thread     |

> | `cub::DeviceReduce ` | -               | +              | ∞           | global memory   |
> 

Reviewer: ```suggestion
| `cub::DeviceReduce ` | -               | +              | ∞           | global memory   |

The details of how each of these layers are implemented is described below. 
```

> ## Thread-scope
> 
> Thread-scope functionality is usually presented with functions and rarely exposed to the user:
> 

Reviewer: Let's avoid any ambiguity with `cuda::thread_scope`s and just avoid the term "scope" all together. 

```suggestion
## Single Thread

Single threaded functionality like `cub::ThreadReduce` is typically implemented as a sequential function and rarely exposed to the user. 

```

```cpp
template <int LENGTH, typename T, typename ReductionOp>
__device__ __forceinline__ 
T ThreadReduce(T* input,         
               ReductionOp reduction_op,  
               T prefix,     

>                Int2Type<LENGTH> /*length*/)

Reviewer: This `Int2Type` type/parameter could use additional explanation if it is a commonly used construct.

Reviewer 2: Maybe we can use the reference-to-array overload of `ThreadReduce` in the example, to avoid having to introduce `Int2Type` here.
```
template <
    int         LENGTH,
    typename    T,
    typename    ReductionOp>
__device__ __forceinline__ T ThreadReduce(
    T           (&input)[LENGTH],
    ReductionOp reduction_op)
{
    return ThreadReduce<LENGTH>((T*) input, reduction_op);
}
```

{
    T retval = prefix;

    #pragma unroll
    for (int i = 0; i < LENGTH; ++i)
        retval = reduction_op(retval, input[i]);

    return retval;
}
```

> ## Warp-scope
> 
> #### Overview
> 
> Warp-scope functionality is provided by structures:

Reviewer: ```suggestion
## Warp-level

CUB warp-level algorithms are specialized for execution by threads in the same CUDA warp. 

These algorithms may only be invoked by `1 <= n <= 32` _consecutive_ threads in the same warp.

#### Overview

Warp-level functionality is provided by types (classes) to provide encapsulation and enable partial template specialization. 

For example, `cub::WarpReduce` is a class template:
```

```cpp
template <typename T, 
          int LOGICAL_WARP_THREADS = 32,
          int LEGACY_PTX_ARCH = 0>
class WarpReduce {
  // ...
  // (1)   define `_TempStorage` type
  // ...
  _TempStorage &temp_storage;
public:
    
  // (2)   wrap `_TempStorage` in uninitialized memory
  struct TempStorage : Uninitialized<_TempStorage> {};

  __device__ __forceinline__ WarpReduce(TempStorage &temp_storage)
  /* (3)   reinterpret cast */
    : temp_storage(temp_storage.Alias()) 
  {}

  /* (4)   actual algorithms */
  __device__ __forceinline__ T Sum(T input);
};
```

> To use warp-scope algorithms, warp size has to be known at compilation time (provided as a template parameter). It's possible to diverge from using the default warp size, in this case:

Reviewer: ```suggestion
In CUDA, the hardware warp size is 32 threads. However, CUB enables warp-level algorithms on "logical" warps of `1 <= n <= 32` threads.

The size of the logical warp is required at compile time via the `LOGICAL_WARP_THREADS` non-type template parameter. This value is defaulted to the hardware warp size of `32`.

There is a vital difference in the behavior of warp-level algorithms that depends on whether `LOGICAL_WARP_THREADS` is a power of two or not:
```

- If `LOGICAL_WARP_THREADS` is a power of two - warp is partitioned into *sub*-warps, each reducing its data independently from other *sub*-warps. The terminology used in CUB: `32` threads are called hardware warp. Groups with less than `32` threads are called *logical* or *virtual* warp since it doesn't correspond directly to any hardware unit.
- If `LOGICAL_WARP_THREADS` is **not** a power of two - there's no partitioning. That is, only the first logical warp executes algorithm.

> 

Reviewer: ```suggestion
<!-- 
TODO: Add George's diagram showing non-power of two logical warps. 
-->
```

      .Reduce(input, valid_items, cub::Sum());
}
```

Due to `LEGACY_PTX_ARCH` issues described above, we can't specialize on the PTX version. `NV_IF_TARGET` shall be used by specializations instead:

```cpp
template <typename T, int LOGICAL_WARP_THREADS, int LEGACY_PTX_ARCH = 0>
struct WarpReduceShfl 
{
   
    
template <typename ReductionOp>
__device__ __forceinline__ T ReduceImpl(T input, int valid_items,
                                        ReductionOp reduction_op)
{
  // ... base case (SM < 80) ...
}

template <class U = T>
__device__ __forceinline__
  typename std::enable_if<std::is_same<int, U>::value ||
                          std::is_same<unsigned int, U>::value,
                          T>::type
    ReduceImpl(T input, 
               int /* valid_items */, 
               cub::Sum /* reduction_op */) 
{
  T output = input;

  NV_IF_TARGET(NV_PROVIDES_SM_80,
               (output = __reduce_add_sync(member_mask, input);),
               (output = ReduceImpl<cub::Sum>(
                    input, LOGICAL_WARP_THREADS, cub::Sum{});));

  return output;
}
    
    
};
```

Specializations are stored in the `cub/warp/specializations` directory.

> #### Temporary storage usage

Reviewer: Given that the specialization section refers to the temporary storage, I think the temporary storage section should be moved above.

> For warp-scope algorithms, it's unsafe to use temporary storage without synchronization:

Reviewer: ```suggestion
Warp-level algorithms require temporary storage for scratch space and inter-thread communication. 

The temporary storage needed for a given instantiation of an algorithm is known at compile time and is exposed through the `TempStorage` member type definition. 

It is the caller's responsibility to create this temporary storage and provide it to the constructor of the algorithm type.

It is possible to reuse the same temporary storage for different algorithm invocations, but it is unsafe to do so without first synchronizing to ensure the first invocation is complete. 

<!-- 
TODO: Add more explanation of the `TempStorage` type and the `Uninitialized` wrapper.
-->

```

> # CUB Developer Overview

Reviewer 3: I would encourage the use of [semantic line breaks](https://ramshankar.org/blog/posts/2019/semantic-line-breaks/) in this document. Some lines are very long and contain multiple sentences. It can be easier to edit and diff Markdown documents when each sentence is written on a separate line. We have adopted this for RAPIDS cuDF documentation and it has been helpful for writing, editing, and reviewing.

Author: Thanks, I didn't know about that approach. It makes sense for version control.

This living document serves as a guide to the design of the internal structure of CUB. 

> The algorithms CUB provides are layered in a way that corresponds to the hierarchy of threads in CUDA where higher-level layers build on top of lower. 

Author: This is @jrhemstad addition, it means that each layer relies on the layer under it. Warp layer relies on the thread layer. Block layer relies on warp layer and finally device layer relies on block layer.

> The algorithms CUB provides are layered in a way that corresponds to the hierarchy of threads in CUDA where higher-level layers build on top of lower. 

Author: @miscco do you have alternative spelling, suggestion?

> The algorithms CUB provides are layered in a way that corresponds to the hierarchy of threads in CUDA where higher-level layers build on top of lower. 

Reviewer: "lower ones" is fine imo.

> The algorithms CUB provides are layered in a way that corresponds to the hierarchy of threads in CUDA where higher-level layers build on top of lower. 

Reviewer: I just rewrote this section to address people's feedback. See https://github.com/NVIDIA/cub/pull/537/files#r931035360

> The algorithms CUB provides are layered in a way that corresponds to the hierarchy of threads in CUDA where higher-level layers build on top of lower. 
> 
> These layers are:
> - Thread 
> - Warp
> - Block
> - Device
> 
> For example, consider `reduce`. CUB has four flavors of `reduce` that correspond to these different layers. Each layer has unique semantics with respect to how they are invoked, how many threads participate, and on which thread(s) the result is valid. 

Reviewer: ```suggestion
CUB provides layered algorithms that correspond to the thread/warp/block/device hierarchy of threads in CUDA. There are distinct algorithms for each layer and higher-level layers build on top of those below. 

For example, CUB has four flavors of `reduce`, one for each layer: `ThreadReduce, WarpReduce, BlockReduce`, and `DeviceReduce`.  Each is unique in how it is invoked, how many threads participate, and on which thread(s) the result is valid. 

These layers naturally build on each other. For example, `WarpReduce` uses `ThreadReduce`, `BlockReduce` uses `WarpReduce`, etc.
```

`cub::ThreadReduce` 
   - A normal function invoked and executed sequentially by a single thread that returns a valid result on that thread
   - Single thread functions are usually an implementation detail and not exposed in CUB's public API

`cub::WarpReduce` and `cub::BlockReduce`
   - A "cooperative" function where threads concurrently invoke the same function to execute parallel work
   - The function's return value is well-defined only on the "first" thread (lowest thread index)

`cub::DeviceReduce` 
   - A normal function invoked by a single thread that spawns additional threads to execute parallel work
   - Result is stored in the pointer provided to the function

>    - Function returns a `cudaError_t` error code

Author: Very good point, thank you @elstehle!

The table below provides a summary of these functions:

| layer                | coop invocation | coop execution | max threads | valid result in |
| -------------------- | --------------- | -------------- | ----------- | --------------- |
| `cub::ThreadReduce`  | -               | -              | 1           | invoking thread |
| `cub::WarpReduce`    | +               | +              | 32          | main thread     |
| `cub::BlockReduce`   | +               | +              | 1024        | main thread     |
| `cub::DeviceReduce ` | -               | +              | ∞           | global memory   |

The details of how each of these layers are implemented is described below. 

> ## Thread-scope

Reviewer: As I was editing each section, I realized it would be helpful to summarize some of the common elements of each layer to avoid having to re-explain it at each layer.

This is fairly basic stuff that will probably end up in the CUB "user" guide, but we can move it once that actually exists :) 

```suggestion
## Common Patterns

While CUB's algorithms are unique at each layer, there are commonalities among all of them:

- Algorithm interfaces are provided as _types_ (classes)
- Algorithms need temporary storage
- Algorithms dispatch to specialized implementations depending on compile-time and runtime information 
- Cooperative algorithms require the number of threads at compile time (template parameter)

Invoking any CUB algorithm follows the same general pattern:
1. Select the class for the desired algorithm
2. Query the temporary storage requirements
3. Allocate the temporary storage
4. Pass the temporary storage to the algorithm 
5. Invoke it via the appropriate member function 

An example of `cub::BlockReduce` demonstrates these patterns in practice:

__global__ void kernel(int* per_block_results) {
  
  // (1) Select the desired class
  // `cub::BlockReduce` is a class template that must be instantiated for the
  // input data type and the number of threads Internally the class is
  // specialized depending on the data type, number of threads, and hardware
  // architecture Type aliases are often used for convenience
  using BlockReduce = cub::BlockReduce<int, 128>;

  // (2) Query the temporary storage
  // The type and amount of temporary storage depends on the selected
  // instantiation
  using TempStorage = typename BlockReduce::TempStorage;

  // (3) Allocate the temporary storage
  __shared__ TempStorage temp_storage;

  // (4) Pass the temporary storage
  // Temporary storage is passed to the constructor of the `BlockReduce` class
  auto block_reduce = BlockReduce{temp_storage};

  // (5) Invoke the algorithm
  // The `Sum()` member function performs the sum reduction of `thread_data`
  // across all 128 threads
  int thread_data[4] = {1, 2, 3, 4};
  int block_result = block_reduce.Sum(thread_data);

  // Alternatively, (4) and (5) are often done at the same time by constructing a temporary instance of the
  // algorithm wrapper type
  int thread_data[4] = {1, 2, 3, 4};
  int result = BlockReduce{temp_storage}.Sum(thread_data);

  per_block_results[blockIdx.x] = result;
}

## Thread-scope
```

Thread-scope functionality is usually presented with functions and rarely exposed to the user:

```cpp
template <int LENGTH, typename T, typename ReductionOp>
__device__ __forceinline__ 
T ThreadReduce(T* input,         
               ReductionOp reduction_op,  
               T prefix,     
               Int2Type<LENGTH> /*length*/)
{
    T retval = prefix;

    #pragma unroll
    for (int i = 0; i < LENGTH; ++i)
        retval = reduction_op(retval, input[i]);

    return retval;
}
```

## Warp-scope

#### Overview

Warp-scope functionality is provided by structures:

```cpp
template <typename T, 
          int LOGICAL_WARP_THREADS = 32,
          int LEGACY_PTX_ARCH = 0>
class WarpReduce {
  // ...
  // (1)   define `_TempStorage` type
  // ...
  _TempStorage &temp_storage;
public:
    
  // (2)   wrap `_TempStorage` in uninitialized memory
  struct TempStorage : Uninitialized<_TempStorage> {};

  __device__ __forceinline__ WarpReduce(TempStorage &temp_storage)
  /* (3)   reinterpret cast */
    : temp_storage(temp_storage.Alias()) 
  {}

  /* (4)   actual algorithms */
  __device__ __forceinline__ T Sum(T input);
};
```

> To use warp-scope algorithms, warp size has to be known at compilation time (provided as a template parameter). It's possible to diverge from using the default warp size, in this case:

Reviewer 4: Can we add a sentence that clearly advices against non power of two threads

Author: Why advice against it if there's a use case?

- If `LOGICAL_WARP_THREADS` is a power of two - warp is partitioned into *sub*-warps, each reducing its data independently from other *sub*-warps. The terminology used in CUB: `32` threads are called hardware warp. Groups with less than `32` threads are called *logical* or *virtual* warp since it doesn't correspond directly to any hardware unit.
- If `LOGICAL_WARP_THREADS` is **not** a power of two - there's no partitioning. That is, only the first logical warp executes algorithm.

It's important to note that `LEGACY_PTX_ARCH` has been recently deprecated. This parameter used to affect specialization selection (see below). It was conflicting with the PTX dispatch refactoring and limited NVHPC support.

#### Specialization

> CUB promises SOL performance when used with NVIDIA hardware. To achieve this, we have to take advantage of hardware-specific features. This is achieved by algorithm specializations. 

Reviewer: ```suggestion
The goal of CUB is to provide users with algorithms that abstract the complexities of achieving speed-of-light performance across a variety of use cases and hardware. 

It is a CUB developer's job to abstract this complexity from the user by providing a uniform interface that statically dispatches to the optimal code path. 

This is usually accomplished via customizing the implementation based on compile time information like the logical warp size, the data type, and the hardware. 

For example, `WarpReduce` dispatches to two different implementations based on if the logical warp size is a power of two (described above):
```

}

template <class U = T>
__device__ __forceinline__
  typename std::enable_if<std::is_same<int, U>::value ||
                          std::is_same<unsigned int, U>::value,
                          T>::type
    ReduceImpl(T input, 
               int /* valid_items */, 
               cub::Sum /* reduction_op */) 
{
  T output = input;

  NV_IF_TARGET(NV_PROVIDES_SM_80,
               (output = __reduce_add_sync(member_mask, input);),
               (output = ReduceImpl<cub::Sum>(
                    input, LOGICAL_WARP_THREADS, cub::Sum{});));

  return output;
}
    
    
};
```

Specializations are stored in the `cub/warp/specializations` directory.

#### Temporary storage usage

For warp-scope algorithms, it's unsafe to use temporary storage without synchronization:

```cpp
using WarpReduce = cub::WarpReduce<int>;

// Allocate WarpReduce shared memory for one warp
__shared__ WarpReduce::TempStorage temp_storage;

int aggregate_1 = WarpReduce(temp_storage).Sum(thread_data_1);
// illegal, has to add `__syncwarp()` between the two
int aggregate_2 = WarpReduce(temp_storage).Sum(thread_data_2);
// illegal, has to add `__syncwarp()` between the two
foo(temp_storage); 
```

> ## Block-scope

Reviewer 4: Ditto the expansion of @jrhemstad wrt to a block overview

CUB_RUNTIME_FUNCTION static 
cudaError_t Sum(
    void *d_temp_storage, size_t &temp_storage_bytes, InputIteratorT d_in,
    OutputIteratorT d_out, int num_items, cudaStream_t stream = 0) 
{
  using OffsetT = int;
  using OutputT =
      cub::detail::non_void_value_t<OutputIteratorT,
                                    cub::detail::value_t<InputIteratorT>>;

  return DispatchReduce<InputIteratorT, OutputIteratorT, OffsetT,
                        cub::Sum>::Dispatch(d_temp_storage, temp_storage_bytes,
                                            d_in, d_out, num_items, cub::Sum(),
                                            OutputT(), stream);
}

};
```

Device-scope facilities always return `cudaError_t` and accept `stream` as the last parameter (main stream by default). First two parameters are always `void *d_temp_storage, size_t &temp_storage_bytes`. The algorithm invocation consists of two phases. During the first phase, temporary storage size is calculated and returned in `size_t &temp_storage_bytes`. During the second phase, `temp_storage_bytes` of memory is expected to be allocated and `d_temp_storage` is expected to be the pointer to this memory.

> > :warning: Even if the algorithm doesn't need shared memory, we still require one byte of memory to be allocated.

Author: I've used the general [term](https://en.wikipedia.org/wiki/Shared_memory#:~:text=In%20computer%20science%2C%20shared%20memory,of%20passing%20data%20between%20programs.) here. I agree, that it might be ambiguous in the CUDA context.

Finally, the tuning looks like:

```cpp
template <typename InputT /* ... */>
struct DeviceReducePolicy 
{
  /// SM35
  struct Policy350 : ChainedPolicy<350, Policy350, Policy300> {
    typedef AgentReducePolicy<256, 20, InputT, 4, BLOCK_REDUCE_WARP_REDUCTIONS,
                              LOAD_LDG>
        ReducePolicy;
  };

  /// SM60
  struct Policy600 : ChainedPolicy<600, Policy600, Policy350> {
    typedef AgentReducePolicy<256, 16, InputT, 4, BLOCK_REDUCE_WARP_REDUCTIONS,
                              LOAD_LDG>
        ReducePolicy;
  };

  /// MaxPolicy
  typedef Policy600 MaxPolicy;
};
```

The kernels in the dispatch layer shouldn't contain a lot of code. Usually, the functionality is extracted into the agent layer. All the kernel does is derive the proper policy type, unwrap the policy to initialize the agent and call one of its `Consume`/`Process` methods. Agents are frequently reused by unrelated device-scope algorithms. 

#### Temporary storage usage

> It's safe to reuse storage in the stream order:

Reviewer 4: That sentence is in contrast to the example, where the last call is illegal. We should make it clear that we need to potentially call `cudaStreamSynchronize(stream)`

Author: The last call is not in stream order, so I don't see any contradiction here
